crates/scroll/chainspec/src/lib.rs

```diff
@@ -0,0 +1,578 @@
+        let genesis_info =
+            scroll_genesis_info.scroll_chain_info.genesis_info.expect("load scroll genesis info");
+
+        // Block-based hardforks
+        let hardfork_opts = [
+            (EthereumHardfork::Homestead.boxed(), genesis.config.homestead_block),
+            (EthereumHardfork::Tangerine.boxed(), genesis.config.eip150_block),
+            (EthereumHardfork::SpuriousDragon.boxed(), genesis.config.eip155_block),
+            (EthereumHardfork::Byzantium.boxed(), genesis.config.byzantium_block),
+            (EthereumHardfork::Constantinople.boxed(), genesis.config.constantinople_block),
+            (EthereumHardfork::Petersburg.boxed(), genesis.config.petersburg_block),
+            (EthereumHardfork::Istanbul.boxed(), genesis.config.istanbul_block),
+            (EthereumHardfork::Berlin.boxed(), genesis.config.berlin_block),
+            (EthereumHardfork::London.boxed(), genesis.config.london_block),
+            (ScrollHardfork::Archimedes.boxed(), genesis_info.archimedes_block),
+            (ScrollHardfork::Bernoulli.boxed(), genesis_info.bernoulli_block),
+            (ScrollHardfork::Curie.boxed(), genesis_info.curie_block),
+        ];
+        let mut block_hardforks = hardfork_opts
+            .into_iter()
+            .filter_map(|(hardfork, opt)| opt.map(|block| (hardfork, ForkCondition::Block(block))))
+            .collect::<Vec<_>>();
+
+        // Time-based hardforks
+        let time_hardfork_opts = [
+            (EthereumHardfork::Shanghai.boxed(), genesis.config.shanghai_time),
+            (ScrollHardfork::Darwin.boxed(), genesis_info.darwin_time),
+            (ScrollHardfork::DarwinV2.boxed(), genesis_info.darwin_v2_time),
+        ];
+
+        let mut time_hardforks = time_hardfork_opts
+            .into_iter()
+            .filter_map(|(hardfork, opt)| {
+                opt.map(|time| (hardfork, ForkCondition::Timestamp(time)))
+            })
+            .collect::<Vec<_>>();
+
+        block_hardforks.append(&mut time_hardforks);
+
+        // Ordered Hardforks
+        let mainnet_hardforks = ScrollHardfork::scroll_mainnet();
+        let mainnet_order = mainnet_hardforks.forks_iter();
+
+        let mut ordered_hardforks = Vec::with_capacity(block_hardforks.len());
+        for (hardfork, _) in mainnet_order {
+            if let Some(pos) = block_hardforks.iter().position(|(e, _)| **e == *hardfork) {
+                ordered_hardforks.push(block_hardforks.remove(pos));
+            }
+        }
+
+        // append the remaining unknown hardforks to ensure we don't filter any out
+        ordered_hardforks.append(&mut block_hardforks);
+
+        Self {
+            inner: ChainSpec {
+                chain: genesis.config.chain_id.into(),
+                genesis,
+                hardforks: ChainHardforks::new(ordered_hardforks),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+struct ScrollGenesisInfo {
```

Comment:
I think a better name for this type is `ScrollConfig`

Reply:
fixed [0cee1ba](https://github.com/scroll-tech/reth/pull/56/commits/0cee1ba025668a3f68e526c138fc08bad4f2aeff)